Create 'edit finder' form: "Filters and options"

ChrisBAshton opened the pull request:
Supersedes #2840. This PR:

- Add summary card for facets, displaying them on the admin page
- Introduces route and form to modify the facets for a finder
- Introduces page to confirm changes to the facets of a finder
- Introduces new "Facet" model to keep logic out of the controller

Trello: https://trello.com/c/KNf2zNFI/3037-create-edit-finder-form-filters-and-options

## Screenshot

Summary page:

![specialist-publisher dev gov uk_admin_cma-cases(iPad Pro)](https://github.com/user-attachments/assets/b7d5568f-0487-40f8-bc01-a8e2386756e9)

Edit facets page:

![Screenshot 2025-01-03 at 16 20 32](https://github.com/user-attachments/assets/91d44e7a-c7fa-44f1-ae03-451296da97c1)

Confirmation page:

![Screenshot 2025-01-03 at 16 21 15](https://github.com/user-attachments/assets/28f3d59c-725f-479f-b73a-37446b3f3924)

With expanded diff:

![Screenshot 2025-01-03 at 16 21 39](https://github.com/user-attachments/assets/ef81ddb2-cf39-463c-a312-5839153437d9)

With expanded schema:

![Screenshot 2025-01-03 at 16 22 02](https://github.com/user-attachments/assets/f9fb9fe3-2fd5-4b10-a22d-a2f1463bf020)

## Caveats

1. The form doesn't work without JS.

2. The question "Do publishers need to select an option...?" in the design is blocked on [moving validation logic into the schema](https://trello.com/c/PArCm42y/3292-dry-up-specialist-publisher-validation), so has been omitted for the time being.

3. All properties [in the finder.jsonnet](https://github.com/alphagov/publishing-api/blob/fb5200249a7e44ac9e6ca84be87ba96586889b16/content_schemas/formats/shared/definitions/finder.jsonnet#L146-L254) are supported, with the exception of the following barely-used filter properties:

- `show_option_select_filter`  ([used only in licence_transactions.json](https://github.com/alphagov/specialist-publisher/blob/fd1c76d40820b3311bf420634cb9615092cb1263/lib/documents/schemas/licence_transactions.json#L56))
- `large` ([used only in licence_transactions.json](https://github.com/alphagov/specialist-publisher/blob/fd1c76d40820b3311bf420634cb9615092cb1263/lib/documents/schemas/licence_transactions.json#L59))
- `open_on_load` ([used only in licence_transactions.json](https://github.com/alphagov/specialist-publisher/blob/fd1c76d40820b3311bf420634cb9615092cb1263/lib/documents/schemas/licence_transactions.json#L60))

---

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

dnkrj committed:
Add summary card for facets

Display facets on the finder admin page.

dnkrj committed:
Add page to edit facets

Introduce route and form to modify the facets for a finder.

We did attempt to have a nice "conditional reveal" on the filter
options for multi/single select, but that doesn't play nicely
with the "Add another" component, so we had to revert it. The
code where we tried it out is in commit:
fcd2e8f45161af9e21618e14f169c906c283b9d8

The other area where the form differs from the design is in the
handling of the 'keys'/'values' (as opposed to 'labels') of the
filter options. The result is the least worst solution to the
problem of: "how do we retain the original filter option key for
_existing_ filter options?".

The design called for a simple textarea comma-separated list of
filter option _labels_ (i.e. "Commercial - fixed wing"), with
nowhere to specify the _key_ / _value_ ("commercial-fixed-wing").
Whilst in that example we can easily convert the label into the
key, there are plenty of examples of existing filter options where
the key is markedly different from the label, and thus submitting
the form would be "lossy" and lead to a larger and unwanted diff.
(Moreover it could lead to real problems if a dev were to go
ahead and merge the diff).

I did explore exposing the key / value pairs in dedicated input
fields, and in theory could then turn the key into an `input
type=hidden` so that it's not exposed to the user. This was
explored in e8810b6a2dc77dbf685aff65bf92f7b11fe6bb36 but the
component doesn't currently allow nesting, so this isn't a
viable way forward for now.

What I've arrived at is I think a good compromise: it is the
label that appears first (followed by the key), and there is no
expectation for the user to edit the key nor to provide a key for
any new filter options they add. The controller automatically
defaults to a kebab-cased key derived from the label, if one is
not provided up front.

(Side note: I couldn't find a Rails native kebabcase method, to
my surprise, so took the regex from
https://jonathanyeong.com/changing-to-kebabcase/).

dnkrj committed:
Add confirmation page for finder facets

Introduces page to confirm changes to the facets of a finder:

- Hides diff / generated schema in a details component.
- Have a shortened summary, don't bother attempting to summarise
  for humans, they can always click on view diff if needed.
- Minimises the diff by sorting the properties as per the most
  common ordering in the existing JSON files (we don't want any
  diffs if someone submits without making any changes!)
- Has summary list facets in the order they're defined (with the
  exception of deleted facets, which are a bit fiddly to slot into
  the correct spot. Jotting them onto the end is sufficient).

ChrisBAshton committed:
Add Diffy dependency and styles

This uses the same dependency, and
[copies the same CSS](https://github.com/alphagov/travel-advice-publisher/blob/767e4308ff2be37ca3497efd43eeb2105699561a/app/assets/stylesheets/diff.scss),
as is used in Travel Advice Publisher.

This will be used on the form confirmation screen in one of the
subsequent commits.

ChrisBAshton committed:
Added Facet model to keep logic outside of AdminController

Following the same pattern as the EmailAlert model introduced in
ae63aa74d931388dd8b874f4461d18c977ff2c61, we've added a Facet
model which provides a translation layer between the form inputs
we expect to add in subsequent commits, and the eventual
representation of the facet in the finder schema JSON.

This logic started off life inline in the AdminController, which
added bloat and complexity to the controller and would have been
extremely difficult to test. The logic is now instead encapsulated
in "Facet" and has an accompanying test suite, so we can be
confident that form inputs are being correctly converted to their
finder schema JSON form.

See next commit for where this class is actually used.

ChrisBAshton committed:
Add feature test for the filters and options form

This integration test checks that the following things work:

- Deleting a facet
- Editing an existing facet
- Adding a new facet

...and that a Zendesk ticket is created.

ryanb-gds commented on app/controllers/admin_controller.rb:
```diff
@@ -75,4 +92,21 @@ def email_alert_params
   end
+
+  def facets_params
+    allowed_facets = %i[
```
Suggesting slightly more specificity - `allowed_facets` to me suggests we're specifying what facets are allowed rather than what attributes a facet can have:
```suggestion
    allowed_facet_params = %i[
```

ryanb-gds commented on app/models/facet.rb:
```diff
@@ -0,0 +1,92 @@
+      elsif str == "false"
+        false
+      end
+    end
```
You could use Active Model for casting ([docs](https://guides.rubyonrails.org/active_model_basics.html#attributes)) if you were up for changing your attribute accessors to full Active Model attribute definitions. I did do this for the finder schema and was a bit equivocal about the outcome, appreciate you may decide/have decided already it's not worth the overhead.

ryanb-gds commented on app/models/facet.rb:
```diff
@@ -0,0 +1,92 @@
+    end
+
+    def facet_allowed_values(values, type)
+      return nil if values.nil? || facet_types_that_allow_enum_values.exclude?(type)
```
I do wonder if we should be using [validation](https://guides.rubyonrails.org/active_model_basics.html#validations) here rather than ignoring invalid values, but can understand why we might choose not to. I can't remember what I did for the email facets in the end - I think I might have been lazy and assumed valid input because of the conditional behaviour making it difficult to provide invalid data, but of course that's not really very good because never trust client input/user could have JS disabled etc.

ChrisBAshton replied:
It's a nice idea. I've now explored it in https://github.com/alphagov/specialist-publisher/commit/d5e35cd275f292739583f7e25d704a5683e943db, but as that commit message explains, it opens a bit of a can of worms - I'm keen that we leave that for a future PR.

ryanb-gds commented on app/views/admin/_facet_fields.html.erb:
```diff
@@ -0,0 +1,110 @@
+  hint: sanitize("Put each option on a new line. The underlying name for existing (live) options will appear in curly braces: please don't edit these, and don't add them for new options (they'll be created automatically later on in the process). Example:<br><strong>Pre-existing value {pre-existing-value}</strong><br><strong>New value</strong>"),
+  name: "facets[#{index}][allowed_values]",
+  rows: 5,
+  value: facet["allowed_values"]&.map { |opt| "#{opt["label"]} {#{opt["value"]}}" }&.join("\n"),
```
What's the value in showing requestors the machine value here? Seems like it would simplify our facet model code to omit it, and they're presumably going to be generated anyway using some kind of snake case/kebab case conversion.

ChrisBAshton replied:
The value is in retaining the _existing_ machine values. There are lots of [examples](https://github.com/alphagov/specialist-publisher/blob/535c7a402b072681a3d3c6838486a9b90dd0901b/lib/documents/schemas/aaib_reports.json#L83-L86) where we can't just kebab-case the name. I go into more detail on the "[Add page to edit facets](https://github.com/alphagov/specialist-publisher/pull/2929/commits/51d2fcd2bc3fc972f8627d8aba1f4930982e0ef0)" commit message.

ryanb-gds replied:
Ah of course, thanks for explaining. Feels a _little_ fragile but I haven't got any better ideas

ryanb-gds commented on app/views/admin/_facets_summary_card.html.erb:
```diff
@@ -0,0 +1,51 @@
+  if (defined? previous_schema)
+    # We're on the edit page, so want to summarise the changes. So overriding `summary_rows`.
+    removed_facets = previous_schema.facets.map { |facet| facet["name"] } - schema.facets.map { |facet| facet["name"] }
+    previous_schema_facet_names = previous_schema.facets.map { |facet| facet["name"] }
+    proposed_schema_facet_names = schema.facets.map { |facet| facet["name"] }
+
+    summary_rows = proposed_schema_facet_names.map do |facet_name|
+      if (previous_schema_facet_names.include?(facet_name))
+        old_facet_config = previous_schema.facets.find { |f| f["name"] == facet_name }
+        new_facet_config = schema.facets.find { |f| f["name"] == facet_name }
+        updated = !facet_name.in?(removed_facets) && !old_facet_config.nil? && old_facet_config != new_facet_config
```
Feels like there ought to be a less verbose way to do this comparison stuff, or maybe we just create a schema diff model and move it elsewhere. Outside the scope of this PR though

ryanb-gds reviewed (commented):
Looking good. A few thoughts, but certainly nothing that would stop this merging, just observations really

ChrisBAshton committed:
Use ActiveRecord to do the boolean casting

As suggested in https://github.com/alphagov/specialist-publisher/pull/2929#discussion_r1901992443

Follows the same pattern as in FinderSchema

ryanb-gds reviewed (approved):
Lovely stuff. Would be good to have one of those fancy JS enabled feature tests for the email subscription options as well.
